File: Remote/Source/Remote Elements/Control State Sets/ControlStateTitleSet.m
//
// ControlStateTitleSet.m
// Remote
//
// Created by Jason Cardwell on 3/26/12.
// Copyright (c) 2012 Moondeer Studios. All rights reserved.
//

#import "ControlStateTitleSet.h"
#import "RemoteElementExportSupportFunctions.h"
#import "RemoteElementImportSupportFunctions.h"
#import "JSONObjectKeys.h"
#import "RemoteElementKeys.h"
//#import "REFont.h"
#import "Remote-Swift.h"

static int ddLogLevel   = LOG_LEVEL_DEBUG;
static int msLogContext = LOG_CONTEXT_CONSOLE;
#pragma unused(ddLogLevel,msLogContext)

@interface ControlStateTitleSet ()
@property (nonatomic) TitleAttributes * normal;
@property (nonatomic) TitleAttributes * disabled;
@property (nonatomic) TitleAttributes * selected;
@property (nonatomic) TitleAttributes * highlighted;
@property (nonatomic) TitleAttributes * highlightedDisabled;
@property (nonatomic) TitleAttributes * highlightedSelected;
@property (nonatomic) TitleAttributes * highlightedSelectedDisabled;
@property (nonatomic) TitleAttributes * selectedDisabled;
@end

@implementation ControlStateTitleSet

@dynamic normal;
@dynamic disabled, selected, selectedDisabled;
@dynamic highlighted, highlightedDisabled, highlightedSelected;
@dynamic highlightedSelectedDisabled;

// @synthesize suppressNormalStateAttributes = _suppressNormalStateAttributes;


////////////////////////////////////////////////////////////////////////////////
#pragma mark Accessors
////////////////////////////////////////////////////////////////////////////////


- (void)setObject:(TitleAttributes *)object forKeyedSubscript:(NSString *)key {
  [super setObject:object forKeyedSubscript:key];
}

- (id)objectForKeyedSubscript:(NSString *)key {

  NSArray * keys = [key keyPathComponents];

  switch ([keys count]) {
    case 2:     // return an attribute value from the attributes of the specified state
    {
      assert(NO);
      return nil;
//      NSString * stateKey = keys[0];
//
//      if ([ControlStateSet validState:stateKey]) {
//        TitleAttributes * titleAttributes = [self valueForKey:stateKey];
//
//        NSString * attributeKey = keys[1];
//
//        if ([[TitleAttributes propertyKeys] containsObject:attributeKey])
//          return [titleAttributes valueForKey:attributeKey];
//
//      }

    }  break;

    case 1:     // return an attributed string with attributes for specified state
      return self[[ControlStateSet stateForProperty:key]];

    default:     // invalid key path
      ThrowInvalidArgument(key, contains illegal key path);
  }

  return nil;
}

- (void)setObject:(TitleAttributes *)object atIndexedSubscript:(NSUInteger)state {
  [super setObject:object atIndexedSubscript:state];
}

- (NSAttributedString *)objectAtIndexedSubscript:(NSUInteger)state {

  NSAttributedString * string = nil;

  TitleAttributes * attributes = [super objectAtIndexedSubscript:state];
  if (attributes) {
    string = (state ? [attributes stringWithFillers:self.normal.attributes] : attributes.string);
  }

  return string;
}


////////////////////////////////////////////////////////////////////////////////
#pragma mark Importing
////////////////////////////////////////////////////////////////////////////////


- (void)updateWithData:(NSDictionary *)data {
  NSManagedObjectContext * moc = self.managedObjectContext;

  [data enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL * stop) {
    if ([ControlStateSet validState:key] && isDictionaryKind(obj))
      self[key] = [TitleAttributes importObjectFromData:obj context:moc];
  }];
}

////////////////////////////////////////////////////////////////////////////////
#pragma mark Exporting
////////////////////////////////////////////////////////////////////////////////


- (MSDictionary *)JSONDictionary {
  MSDictionary * dictionary = [super JSONDictionary];

  // remove entries for state dictionaries
  for (NSString * key in [dictionary copy])
    if ([ControlStateSet validState:[key keyPathComponents][0]])
      [dictionary removeObjectForKey:key];




  for (NSString * key in [ControlStateSet validProperties])
    SafeSetValueForKey(((TitleAttributes *)[self valueForKey:key]).JSONDictionary, key, dictionary);

  [dictionary compact];
  [dictionary compress];

  return dictionary;
}

@end

File: Remote/Source/Remote Elements/Control State Sets/ControlStateTitleSet.swift
//
//  ControlStateTitleSet.swift
//  Remote
//
//  Created by Jason Cardwell on 11/22/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

import Foundation
import UIKit
import CoreData
import MoonKit

extension UIControlState: JSONValueConvertible {
  public var JSONValue: String {
    var flags: [String] = []
    if self & UIControlState.Highlighted != nil { flags.append("highlighted") }
    if self & UIControlState.Selected    != nil { flags.append("selected")    }
    if self & UIControlState.Disabled    != nil { flags.append("disabled")    }
    if flags.count == 0 { flags.append("normal") }
    return " ".join(flags)
  }
  public init?(JSONValue: String) {
    let flags = split(JSONValue){$0 == " "}
    if !contains(1...3, flags.count) { return nil }
    var state = UIControlState.Normal
    for flag in flags {
      switch flag {
        case "highlighted": state |= UIControlState.Highlighted
        case "selected":    state |= UIControlState.Selected
        case "disabled":    state |= UIControlState.Disabled
        case "normal":      if state != UIControlState.Normal { return nil }
        default:            return nil
      }
    }
    self = state
  }
}

extension UIControlState: EnumerableType {
  public static var all: [UIControlState] {
    return [.Normal, .Highlighted, .Selected, .Disabled,
      .Highlighted | .Selected, .Highlighted | .Disabled,
      .Highlighted | .Selected | .Disabled]
  }
  public static func enumerate(block: (UIControlState) -> Void) { apply(all, block) }
}

@objc(ControlStateTitleSet)
class ControlStateTitleSet: ControlStateSet {

  @NSManaged var disabled:                    DictionaryStorage?
  @NSManaged var selectedDisabled:            DictionaryStorage?
  @NSManaged var highlighted:                 DictionaryStorage?
  @NSManaged var highlightedDisabled:         DictionaryStorage?
  @NSManaged var highlightedSelected:         DictionaryStorage?
  @NSManaged var normal:                      DictionaryStorage?
  @NSManaged var selected:                    DictionaryStorage?
  @NSManaged var highlightedSelectedDisabled: DictionaryStorage?

  /**
  setTitleAttributes:forState:

  :param: attributes TitleAttributes?
  :param: state UIControlState
  */
  func setTitleAttributes(attributes: TitleAttributes?, forState state: UIControlState) {
    var property: String?
    switch state {
      case UIControlState.Normal:                                                      property = "normal"
      case UIControlState.Highlighted:                                                 property = "highlighted"
      case UIControlState.Selected:                                                    property = "selected"
      case UIControlState.Disabled:                                                    property = "disabled"
      case UIControlState.Highlighted|UIControlState.Selected:                         property = "highlightedSelected"
      case UIControlState.Highlighted|UIControlState.Disabled:                         property = "highlightedDisabled"
      case UIControlState.Selected|UIControlState.Disabled:                            property = "selectedDisabled"
      case UIControlState.Highlighted|UIControlState.Selected|UIControlState.Disabled: property = "highlightedSelectedDisabled"
      default:                                                                         break
    }

    var storage: DictionaryStorage?
    if property != nil {
      if attributes == nil { setValue(nil, forKey: property!) }
      else {
        storage = valueForKey(property!) as? DictionaryStorage
        if storage == nil {
          storage = DictionaryStorage(context: managedObjectContext)
          setValue(storage, forKey: property!)
        }
        assert(storage != nil, "what happened? we should have created storage if it didn't exist")
        storage?.dictionary = attributes!.dictionaryValue
      }
    }

  }

  /**
  titleAttributesForState:

  :param: state UIControlState

  :returns: TitleAttributes?
  */
  func titleAttributesForState(state: UIControlState) -> TitleAttributes? {
    var property: String?
    switch state {
      case UIControlState.Normal:                                                      property = "normal"
      case UIControlState.Highlighted:                                                 property = "highlighted"
      case UIControlState.Selected:                                                    property = "selected"
      case UIControlState.Disabled:                                                    property = "disabled"
      case UIControlState.Highlighted|UIControlState.Selected:                         property = "highlightedSelected"
      case UIControlState.Highlighted|UIControlState.Disabled:                         property = "highlightedDisabled"
      case UIControlState.Selected|UIControlState.Disabled:                            property = "selectedDisabled"
      case UIControlState.Highlighted|UIControlState.Selected|UIControlState.Disabled: property = "highlightedSelectedDisabled"
      default:                                                                         break
    }
    var storage: DictionaryStorage?
    if property != nil { storage = valueForKey(property!) as? DictionaryStorage }
    return storage == nil ? nil : TitleAttributes(storage: storage!.dictionary as [String:AnyObject])
  }

  func attributedStringForState(state: UIControlState) -> NSAttributedString? {
    var string: NSAttributedString?
    if let indexedAttributes = self[state.rawValue] as? DictionaryStorage {
      let attributes = TitleAttributes(storage: indexedAttributes.dictionary as [String:AnyObject])
      if state == UIControlState.Normal {
        string = attributes.string
      } else {
        var normalAttributes: TitleAttributes?
        if normal != nil { normalAttributes = TitleAttributes(storage: normal!.dictionary as [String:AnyObject]) }
        string = normalAttributes == nil ? attributes.string : attributes.stringWithFillers(normalAttributes!.attributes)
      }
    }
    return string
  }

  /**
  updateWithData:

  :param: data [NSObject AnyObject]
  */
  override func updateWithData(data: [NSObject:AnyObject]) {
    super.updateWithData(data)

    if let jsonData = data as? [String:[String:AnyObject]] {
      for (stateKey, dictionary) in jsonData {
        if let controlState = UIControlState(JSONValue: stateKey) {
          setTitleAttributes(TitleAttributes(JSONValue: dictionary), forState: controlState)
        }
      }
    }
  }

  /**
  JSONDictionary

  :returns: MSDictionary
  */
  override func JSONDictionary() -> MSDictionary {
    let dictionary = super.JSONDictionary()
    UIControlState.enumerate {
      if let attributes = self.titleAttributesForState($0) { dictionary[$0.JSONValue] = attributes.JSONValue }
      else { dictionary.removeObjectForKey($0.JSONValue) }
    }
    return dictionary
  }

}
